### DISCUSSION TITLE
C/C++ interop

### DISCUSSION THREAD

**Discussion by mateodif** [Zelex/jo_lisp#9] [Ideas]
One of Clojure's main benefits include being able to interop with the native underlying environment that it's running on: Clojure interops with the JVM, Clojure CLR with .NET, ClojureScript with JavaScript and so on.
It would be very interesting to be able to interop with C++.

[Jank](https://github.com/jank-lang/jank) has C++ interop that looks like this:
```clojure
(defn string? [o]
  (native/raw "__value = make_box(#{ o }#->as_string() != nullptr)"))
```
You can read more about it [here](https://github.com/jank-lang/jank/blob/main/DESIGN.md#interop).

**Comment by Zelex** [comment]
Was planning on doing this first through native library hooks. dlopen, dlsym, dlclose etc hooks. Its a ways off though. 

Then later theoretically you could write C/C++ code in your clj itself and it would compile it to a .so file and use the above methods to automatically hook into it. 

This is all in theory of course.

**Comment by Zelex** [comment]
Something like 

(native/c++ 
  int foo(int bar) { 
    return bar*2;
  }
)

Is possible, but I'd have to build a C++ parser into the program, which I really don't want to do. 

Putting the code inside a string is icky too because what if you need a string inside that string, then your escaping all over and f' that. 

You could do something like C++ Raw strings, which is probably the more reasonable approach.
